proposal/package-importer.d.ts.md

```diff
@@ -0,0 +1,527 @@
+  each `subpathVariants` as `subpathVariant`.
+
+  > The PACKAGE_EXPORTS_RESOLVE algorithm always includes a `default` condition,
+  > so one does not have to be passed here.
+
+* If `resolvedPaths` contains more than one resolved URL, throw an error.
+
+* If `resolvedPaths` contains exactly one resolved URL, return it.
+
+* If `subpath` has an extension, return null.
+
+* Let `subpathIndex` be `subpath` + `"/index"`.
+
+* Let `subpathIndexVariants` be the result of [Export load paths] with `subpathIndex`.
+
+* Let `resolvedIndexPaths` be a list of the results of calling
+  `PACKAGE_EXPORTS_RESOLVE(packageRoot, subpathVariant, exports, [condition])`
+  as defined in the [Node resolution algorithm specification], with each
+  `subpathIndexVariants` as `subpathVariant`.
+
+* If `resolvedIndexPaths` contains more than one resolved URL, throw an error.
+
+* If `resolvedIndexPaths` contains exactly one resolved URL, return it.
+
+* Return null.
+
+> Where possible in Node, implementations can use [resolve.exports] which
+> exposes the Node resolution algorithm, allowing for per-path custom
+> conditions, and without needing filesystem access.
+
+[Export load paths]: #export-load-paths
+[resolve.exports]: https://github.com/lukeed/resolve.exports
+
+### Resolving package root values
+
+This algorithm takes a string `packagePath`, which is the root directory for a
+package, and `packageManifest`, which is the contents of that package's
+`package.json` file, and returns a file URL.
+
+* Let `sassValue` be the value of `sass` in `packageManifest`.
+
+* If `sassValue` is a relative path with an extension of `sass`, `scss` or
+  `css`:
+
+  * Return the canonicalized `file:` URL for `${packagePath}/${sassValue}`.
+
+* Let `styleValue` be the value of `style` in `packageManifest`.
+
+* If `styleValue` is a relative path with an extension of `sass`, `scss` or
+  `css`:
+
+  * Return the canonicalized `file:` URL for `${packagePath}/${styleValue}`.
+
+* Otherwise return the result of [resolving a `file:` URL for extensions] with
+  `packagePath + "/index"`.
+
+[resolving a `file:` URL for extensions]: ../spec/modules.md#resolving-a-file-url-for-extensions
+
+[URL path segments]: https://url.spec.whatwg.org/#url-path-segment
+
+### Export Load Paths
+
+This algorithm takes a relative URL path `subpath` and returns a list of
+potential subpaths, resolving for partials and file extensions.
+
+* Let `paths` be a list.
+
+* If `subpath` ends in `.scss`, `.sass`, or `.css`:
+
+  * Add `subpath` to `paths`.
+
+* Otherwise, add `subpath` + `.scss`, `subpath` + `.sass`, and `subpath` +
+  `.css` to `paths`.
+
+* If `subpath`'s [basename] does not start with `_`, for each `item` in
+  `paths`, prepend `"_"` to the basename, and add to `paths`.
+
+* Return `paths`.
+
+[basename]: ../spec/modules.md#basename
+
+## Embedded Protocol
+
+```proto
+message CompileRequest {
+  message Importer {
+    PackageImporter packageImporter = 14;
```

Comment:
This should be a new member of the `Importer.importer` oneof, since it's a new type of importer. Also, it doesn't need index 14; the index just needs to be unique within the `Importer` record.

Reply:
Ah, right, I forgot to update to mirror the move in the JS API from a flag to a Importer object.

I updated this in https://github.com/sass/sass/pull/3660/commits/8658bc3e29f34bfbe206cac35b2f24c44ae90ec4. 

For the `entryPointUrl`, I added a NodePackageImporter message with a string `entry_point_url`. The other option I considered but rejected was flattening the object with a single key into-

```proto
    oneof importer {
      string node_package_importer_entry_point_url = 4;
    }
```

The nested message style I went with would allow for easier future additions, but I'm fine with either option (or something else).

@ntkme  Thanks for catching and explaining this- does this address the need?